spec/client/ics-028-wasm-client/README.md

```diff
@@ -0,0 +1,451 @@
+author: Parth Desai <parth@chorus.one>
+created: 2020-10-13
+modified: 2020-10-13
+---
+
+## Synopsis
+
+This specification document describes an interface to a client (verification algorithm) stored as a Wasm bytecode for a blockchain.
+
+### Motivation
+
+Wasm based light clients are decoupled from SDK source code, which enables one to upgrade existing light client or add support for other blockchain without modifying SDK source code.
+
+### Definitions
+
+Functions & terms are as defined in [ICS 2](../ics-002-client-semantics).
+
+`currentTimestamp` is as defined in [ICS 24](../ics-024-host-requirements).
+
+`Wasm Client Code` refers to Wasm bytecode stored in the client store, which provides a target blockchain specific implementation of [ICS 2](../ics-002-client-semantics).
+
+`Wasm Client` refers to a particular instance of `Wasm Client Code` defined as a tuple `(Wasm Client Code, ClientID)`.
+
+`Wasm VM` refers to a virtual machine capable of executing valid Wasm bytecode.
+
+
+### Desired Properties
+
+This specification must satisfy the client interface defined in ICS 2.
+
+## Technical Specification
+
+This specification depends on correctness of the `Wasm Client code` in context of consensus algorithm of its target `blockchain`, as well as correct instantiation of `Wasm Client`.
+
+### Client state
+
+The Wasm client state tracks location of the Wasm bytecode via `codeId`. Binary data represented by `data` field is opaque and only interpreted by the Wasm Client Code. `type` represents client type.
+`type` and `codeId` both are immutable.
+
+```typescript
+interface ClientState {
+  codeId: []byte
+  data: []byte
+  frozen: boolean
+  frozen_height: Height
+  latest_height: Height
+  type: String
+}
+```
+
+### Consensus state
+
+The Wasm consensus state tracks the timestamp (block time), `Wasm Client code` specific fields and commitment root for all previously verified consensus states.
+`type` and `codeId` both are immutable.
+
+```typescript
+interface ConsensusState {
+  codeId: []byte
+  data: []byte
+  timestamp: uint64
+  root: MerkleRoot
+  type: String
+}
+```
+
+### Height
+
+The height of a Wasm light client instance consists of two `uint64`s: the epoch number, and the height in the epoch.
+
+```typescript
+interface Height {
+  epochNumber: uint64
+  epochHeight: uint64
+}
+```
+
+Comparison between heights is implemented as follows:
+
+```typescript
+function compare(a: Height, b: Height): Ord {
+  if (a.epochNumber < b.epochNumber)
+    return LT
+  else if (a.epochNumber === b.epochNumber)
+    if (a.epochHeight < b.epochHeight)
+      return LT
+    else if (a.epochHeight === b.epochHeight)
+      return EQ
+  return GT
+}
+```
+
+This is designed to allow the height to reset to `0` while the epoch number increases by one in order to preserve timeouts through zero-height upgrades.
+
+### Headers
+
+Contents of Wasm client headers depend upon `Wasm Client Code`.
+
+```typescript
+interface Header {
+  data: []byte
```

Comment:
Methods that don't have access to storage

### Header ([link](https://github.com/cosmos/ibc-go/blob/main/modules/core/exported/client.go#L210)

```
ClientType() string
GetHeight() Height
ValidateBasic() error
```

This definition looks stale. It should be:
```
interface Header { 
   data: []byte,
   height: Height, // required by GetHeight() 
}
```

Reply:
ditto on ValidateBasic

I'm a bit hesitant to make `GetHeight` stateful here since it gets used in ValidateBasic checks and without it the ValidateBasic isn't doing much at all. If you could clarify why you need access to storage, and the difficulties with keeping it stateless that would be helpful.

Reply:
@AdityaSripal sorry I phrased it wrongly. The storage is not needed here, but I wanted to give some context as to why the `height: Height` is a separate field (not a part of `data`), this is [during the UpdateClient call the `GetHeight()` is required](https://github.com/cosmos/ibc-go/blob/1ff0da6694d08e494a9c87d5b9737ab5312677a0/modules/core/02-client/keeper/client.go#L81) to set the consensus state.

That's the only significant use case of `GetHeight` I found

Reply:
updated `Header` [here](https://github.com/ParthDesai/ibc/commit/d7c33fc5c01ff4c07000f89ff9da04c31c27ed23#diff-081ac6bcbed0e39a406e9c83aa79a058605a646a4047d5be67cb84dd9b83877bL106)

Reply:
Sorry to interject without reading through all the context, but I think `GetHeight()` will be removed soonish. In a couple months I'd like to start work on #569 which would move setting of IBC client information to light client implementations. This would mean the consensus state would be set by the light client, which could directly cast to its Header definition as opposed to using the header interface. The header interface would only be really used to pass the header from the user message to 02-client and finally to the IBC client (which would not require `GetHeight()` usage

I'm not really sure what this means for the decision being discussed, but whatever is decided should be fine since it'll likely be short lived. It sounds like `GetHeight` will still be stateless in the short term? The `height` field would just be removed from data once `GetHeight` is removed